docs/cvl/expr.md

```diff
@@ -359,3 +366,92 @@ state `s`.
+3. The state of all ghost variables and functions
+
+Thus, if any field in any contract's storage differs between `s1` and `s2`, the expression will return `false`.
+The expression `s1 != s2` behaves similarly, except it will return `false` if all contract storage, balances,
+and ghosts are equal between `s1` and `s2`.
+
+Storage comparisons also support narrowing the scope of comparison to specific components of the global
+state represented by `storage` variables. This syntax is `s1[r] storage_relop s2[r]`, where `r` is a "storage comparison basis", and `storage_relop` is either `!=` or `==`. The valid bases of comparison are:
+
+1. The name of a contract imported with a {ref}`using statement <using-stmt>`,
+2. The keyword `nativeBalances`, or
+3. The name of a ghost variable or function
+
+It is an error to use different bases on different sides of the comparison operator, and it is also
+an error to use a comparison basis on one side and not the other.
+The application of the basis restricts the comparison
+to only consider the portion of global state identified by the basis.
+
+If the qualifier is a contract identifier
+imported via `using`, then the comparison operation will only consider the storage fields of that contract. For example:
+
+```cvl
+using MyContract as c;
+using OtherContract as o;
+
+rule compare_state_of_c(env e) {
+   storage init = lastStorage;
+   o.mutateOtherState(e);
+   assert lastStorage[o] == init[o];
+}
+```
+
+will pass verification whereas:
+
+```cvl
+using MyContract as c;
+using OtherContract as o;
+
+rule compare_state_of_c(env e) {
+   storage init = lastStorage;
+   c.mutateContractState(e);
+   assert lastStorage[c] == init[c];
+}
+```
+
+will not. Note comparing contract's state using this method will **not** compare the balance of the contract between the
+two states.
+
+If the qualifier is the identifier `nativeBalances`, then the account balances
+of all contracts are compared between the two storage states. 
+Finally, if the basis is the name of a ghost function or variable, the value of that
+function/variable are compared between storage states.
+
+Two ghost functions are considered equal if they have the same outputs for all input arguments.
+
+```{warning}
+The default behavior of the Prover on unresolved external calls is to pessimistically havoc contract
+state and balances. This behavior will render most storage comparisons that incorporate such
+state useless. Care should be taken (using {ref}`summarization <summaries>`) to ensure that rules
+that compare storage do not encounter this behavior.
+```
+
+```{warning}
+The grammar admits storage comparisons for both equality and inequality
+that occur arbitrarily nested within expressions. However, support within the Prover for
+these comparisons is primarily aimed at assertions of storage equality, e.g., `assert s1 == s2`.
+Support for storage inequality _and_ nesting comparisons within other expressions is considered
+experimental.
+```
```

Comment:
Do we have known bugs here, or is this just a warning that this isn't fully tested yet (or that there are specific things we won't support?)  I'd be in favor of removing this warning.  Otherwise, I think it should be a bit clearer about what is / isn't supported.

Reply:
Also, I'm not sure whether "storage inequality _and_ nesting comparisons" means the combination of those two things or each of them individually

Reply:
well, they use quantifiers, which means that we are entirely best effort.